### PR TITLE
Fixes to example/stateful_lstm.py

### PR DESCRIPTION
**Pull request by shadiakiki1986**
- rename file to `lstm_stateful.py` so it shows up together with the other `lstm_*.py` example files
- fix input average: the existing code is not doing any averaging because `cos[i + 1:i + lahead + 1]` is just one point. I think the purpose was for it to be a 2-point average.
- aesthetic progress fix: the progress message was 0-based, i.e. first message was `Epoch 0 / 25`.

### CODE REVIEW

**Comment by shadiakiki1986** [comment]
Btw I don't think the cosine is a good example, even with the 2-point averaging.
Even if the LSTM only learns to use the current value as a prediction,
(instead of learning to average the two points),
its mean squared error would be very low.
A better example would be to replace

    cos = gen_cosine_amp()

with something like

    cos = np.random.uniform(-100, +100, (int(50e3), 1, 1))

In this case, the stateful LSTM reaches

    Epoch 25 / 25 - 50000/50000 - 4s - loss: 835.5292

whereas the stateless counterpart can reach

    Epoch 4/25 - 25000/25000 - 3s - loss: 8.4869     

I think this shows that maybe something needs to be fixed in this model with this particular example

**Comment by fchollet** [comment]
> Btw I don't think the cosine is a good example, even with the 2-point averaging.

I agree. Feel free to replace this example script with something new entirely, that would do a better job at demonstrating stateful RNNs.

**Comment by shadiakiki1986** [comment]
So I replaced the cosine with random uniform data.
I also added a stateful model to show a comparison of performance between the two.
Now this example demonstrates a case that stateful can converge to a good fit whereas stateless cannot.

**Comment by shadiakiki1986** [comment]
I also modified the code again so that it is possible that
the user modify the `lahead` and `tsteps` in the script head
in order to make the stateless LSTM.

For example, with `tsteps=2` (i.e. 2-point average) and `lahead=1` (i.e. LSTM not seeing except 1 point), the stateful LSTM can converge whereas the stateless LSTM cannot.
On the other hand, with `tsteps=2` and `lahead=2` (i.e. LSTM sees 2 points), both stateful and stateless LSTMs can converge.

@fchollet what do you think?

File: examples/lstm_stateful.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+
+def create_model(stateful: bool):
+    model = Sequential()
+    model.add(LSTM(20,
+              input_shape=(lahead, 1),
+              batch_size=batch_size,
+              return_sequences=False,
+              stateful=stateful,
+              activation='tanh'))
```
Comment (fchollet) [inline examples/lstm_stateful.py]: You can remove `activation` and `return_sequences` since they take the default values.

File: examples/lstm_stateful.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+# length of input
+input_len = 1000
+
+# training parameters
+batch_size = 1
+epochs = 10
+
+# ----------------------------
+# DO NOT EDIT UNDER THIS LINE
+# ----------------------------
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Example scripts are meant to be edited, please remove.

File: examples/lstm_stateful.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+model_stateful = create_model(stateful=True)
+
+
+# split train/test data
+def split_data(X, y, ratio: int = 0.8):
+    to_train = int(input_len * ratio)
+    # tweak to match with batch_size
+    to_train -= to_train % batch_size
+
+    X_train = X[:to_train]
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Please use lowercase variable names.

**Comment by shadiakiki1986** [comment]
Done. Instead of removing "DO NOT EDIT", I replaced it with "MAIN PROGRAM"

File: examples/lstm_stateful.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+import pandas as pd
+
+# -------------------
+# EDITABLE PARAMETERS
+# -------------------
+
+# timesteps to use in the output averaging of the input
+# e.g. use 2 for 2-point average
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Please add more info, it is not possible to understand this comment without more context.

File: examples/lstm_stateful.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+tsteps = 2
+
+# number of elements ahead that are used to make the LSTM prediction
+# when lahead >= tsteps, the stateless LSTM can converge
+# similar to the stateful LSTM
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Same, please add more info. Maybe a concrete example would help.

File: examples/lstm_stateful.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
+'''Example script showing how to use stateful RNNs
+to model long sequences efficiently.
+Depending on the parameters, either only the stateful LSTM converges,
+or both the stateful and stateless LSTM converge.
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Please flesh out the docstring by explaining what the example is about: what the task is, what the data look like, how the model is set up. Users should be able to go into the code already understanding what they're going to be looking at.

File: examples/lstm_stateful.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+# ------------
+# MAIN PROGRAM
+# ------------
+
+print("*" * 33)
+if lahead >= tsteps:
+    print("STATELESS LSTM WILL ALSO CONVERGE")
+else:
+    print("STATELESS LSTM WILL NOT  CONVERGE")
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Typo: two spaces after NOT

File: examples/lstm_stateful.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+print("*" * 33)
+
+# add a few points to account for the
+# nan-values that will be dropped in preprocessing
```
Comment (fchollet) [inline examples/lstm_stateful.py]: What does this mean?

File: examples/lstm_stateful.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+print('Plotting input and expected output')
+plt.plot(data_input[0][:10], '.')
+plt.plot(expected_output[0][:10], '-')
+plt.legend(['Input', 'Expected output'])
+plt.title('Input')
+plt.show()
+
+
+from keras.models import Sequential
+from keras.layers import Dense, LSTM
```
Comment (fchollet) [inline examples/lstm_stateful.py]: Put the imports at the top of the file.

File: examples/lstm_stateful.py (hunk 8 above)
Comment (shadiakiki1986) [reply examples/lstm_stateful.py]: I extended the comment to explain as follows:

Since the output is a moving average of the input,
the first few points of output will be NaN
and will be dropped from the generated data
before training the LSTM.
Also, when lahead > 1, the preprocessing step later of "rolling window view"
will also cause some points to be lost.
For aesthetic reasons,
in order to maintain generated data length = input_len after pre-processing,
add a few points to account for the values that will be lost.

**Comment by shadiakiki1986** [comment]
Done
